src/components/UI/EditableValueLabel.tsx

```diff
@@ -0,0 +1,199 @@
+  validationFunc,
+  value,
+}: IEditableValueLabel) => {
+  const valueElement = useRef<HTMLSpanElement>(null);
+  const inputElement = useRef<HTMLInputElement>(null);
+
+  const [editing, setEditing] = useState(isNew);
+  const [
+    { internalValue, isValid, validationError },
+    setInternalValue,
+  ] = useReducer<Reducer<IReducerState, string>>(
+    (_, newInternalValue) => ({
+      internalValue: newInternalValue,
+      ...validationFunc(newInternalValue),
+    }),
+    {
+      internalValue: value,
+      ...validationFunc(value),
+    }
+  );
+
+  const keyHandler: KeyboardEventHandler<HTMLInputElement> = ({ key }) => {
+    if (key === 'Escape') {
+      setInternalValue(value);
+      onCancel();
+      setEditing(false);
+    } else if (isValid === true && key === 'Enter') {
+      onSave(internalValue);
+      setEditing(false);
+    }
+  };
+
+  return (
+    <Editable
+      allowEdit={allowEdit}
+      ref={valueElement}
+      onClick={() => {
+        if (allowEdit) {
+          onEdit();
+          setEditing(true);
+        }
+      }}
+    >
+      {value}
+      <Overlay
+        // @ts-ignore
+        target={valueElement.current}
+        placement='top'
+        show={editing}
+        shouldUpdatePosition={true}
+        animation={false}
+      >
+        <EditValueTooltip id={`edit-${value}`}>
+          <EditModeWrapper>
+            <ValueInput
+              ref={inputElement}
+              type='text'
+              onChange={({ target: { value: newRawValue } }) =>
+                setInternalValue(newRawValue)
+              }
+              value={internalValue}
+              onKeyUp={keyHandler}
+            />
+            <Overlay
+              // @ts-ignore
+              target={inputElement.current}
+              placement='top'
+              show={!isValid}
+              shouldUpdatePosition={true}
+              animation={false}
+            >
+              <ValidationErrorTooltip id={`edit-${value}-error`}>
+                {validationError}
+              </ValidationErrorTooltip>
+            </Overlay>
+            <Button
+              bsStyle='success'
+              disabled={!isValid}
+              onClick={() => {
+                onSave(internalValue);
+                setEditing(false);
+              }}
+            >
+              Save
+            </Button>
+            <Button
+              bsStyle='link'
+              onClick={() => {
+                setInternalValue(value);
+                onCancel();
+                setEditing(false);
+              }}
+            >
+              Cancel
+            </Button>
+          </EditModeWrapper>
+        </EditValueTooltip>
+      </Overlay>
+    </Editable>
+  );
+};
+
+EditableValueLabel.propTypes = {
+  allowEdit: PropTypes.bool,
+  isNew: PropTypes.bool,
+  onCancel: PropTypes.func,
+  onEdit: PropTypes.func,
+  onSave: PropTypes.func,
+  validationFunc: PropTypes.func,
+  value: PropTypes.string.isRequired,
```

Comment:
What do you think about always having required props first, and then the optional ones? (same as in the props interface)

Reply:
I've sorted by first letter, but I like your idea more :+1: